### DISCUSSION TITLE
ImageUtilities Fails to Copy

### DISCUSSION THREAD

**Discussion by KingGamingYT** [tsukasa/BdBrowser#45] [Problems and issues]
Hi!
A very peculiar bug I've encountered is that, when using BdBrowser, the plugin ImageUtilities' context menu buttons that it adds for copying media stop working (they copy nothing to the clipboard). In addition, trying to copy an image using the "Copy Image" option will throw an error in the console. I've attached both the menu options and the error below. I'm unsure if this is an ImageUtilities issue, or a wider issue with the plugin's library, BDFDB, since plugins which use that library are usually heavily integrated with it.
![image](https://github.com/user-attachments/assets/1480bf23-9490-42b8-8838-9af0af348fa7)
![image](https://github.com/user-attachments/assets/b4adf78f-aaf0-4ab9-9128-bfa966bafd2d)

**Comment by tsukasa** [comment]
Greetings,

thank you for the report! Discord has seemingly started to litter its codebase with assertions like the one from your screenshot. The assertion is being hit if a platform-related module (`BdApi.Webpack.getAllByKeys("isPlatformEmbedded");`) reports that you are running the web client.

The variables for this module are initialized exactly once during the client's startup/initialization and stay persistent for the rest of the session.

Moving certain shims into the preloader causes Discord to perceive the web client as a native client. However, this also causes new issues that break the entire web client due to additional missing dependencies.

Analysis is still ongoing, however this is the underlying cause.

Hope you'll have a great new year!

**Reply by tsukasa** [reply]
Following up with a more complete view of the underlying complexity: Discord's chosen approach here is something I consider unfixable without investing a ton of time into emulating a lot more  Node features and essentially turning the web client into the desktop client.

This is unlikely to happen, given the necessary amount of work and the likely instability/compatibility issues this causes vs. the existence of alternative solutions like Vencord or simply modifying problematic plugins to use browser features that achieve the same (as an example).

If someone has a ton of time and patience to move from exception to exception to even just make the client start again without crashing, I welcome PRs or forks that implements the missing features. The unfortunate reality is that several known bugs and nice-to-have ideas in BdBrowser today stay unresolved/unimplemented due to time constraints, so a de-facto version 2 that requires a ton more base features seems unlikely.

If Discord's codebase should move more into this direction, these types of issues will most likely fall under "wontfix", with a strong recommendation for alternative, made-for-web solutions.

**Reply by KingGamingYT** [reply]
Ah, I see. Thank you for the thorough explanation! Luckily, because it's a browser, clicking into the image viewer and _then_ right clicking the image will give you the browser's native context menu, allowing you to copy the media with no issue.
